pylint/checkers/similar.py

```diff
@@ -598,10 +599,18 @@ def stripped_lines(
     if ignore_imports or ignore_signatures:
         tree = astroid.parse("".join(lines))
     if ignore_imports:
-        node_is_import_by_lineno = (
-            (node.lineno, isinstance(node, (nodes.Import, nodes.ImportFrom)))
-            for node in tree.body
-        )
+
+        def _get_imports(root: list[nodes.NodeNG]) -> Iterable[tuple[int, bool]]:
+            """Find import nodes under given node."""
+            for node in root:
+                if isinstance(node, astroid.If) and isinstance(
+                    safe_infer(node.test), nodes.Const
+                ):
+                    # Check for imports in a body of a constant condition
+                    yield from _get_imports(node.body)
+                yield node.lineno, isinstance(node, (nodes.Import, nodes.ImportFrom))
+
+        node_is_import_by_lineno = _get_imports(tree.body)
         line_begins_import = {
             lineno: all(is_import for _, is_import in node_is_import_group)
             for lineno, node_is_import_group in groupby(
```

Comment:
```suggestion
        line_begins_import = {
            node.lineno: True
            for node in tree.nodes_of_class((nodes.Import, nodes.ImportFrom))
        }
```

Reply:
Thank you. Regarding the inference, we can easily drop it in the line 607 and just assume any constant and non-constant condition. However, loops containing imports will be still ignored.

Unfortunately, placing only import nodes to `line_begins_import` in the suggestion breaks the detection loop because it checks topmost nodes (`tree.body`) (like `if` statement in this test case) in `line_begins_import` dict..

So the case like:
```py
import typing

if typing.TYPE_CHECKING:
    import os
    import sys

if typing.TYPE_CHECKING:
    x = 1
    y = 2
    import math
```
don't spot the duplicate code for assignments.

I think, the test case `test_no_hide_code_with_imports` can't be handled properly because existing code works on line-level not node-level. Therefore, all these lines do not start with imports and not skipped from similarity check.

Reply:
Hi, thanks for the quick reply.

I'm not totally following, but it could be a gap in my own understanding.

> don't spot the duplicate code for assignments.

This is a good test case, but it's also failing with the current PR, no? It passes with my suggestion when adding an appropriately small ``--max-similarity-lines``.

> Unfortunately, placing only import nodes to line_begins_import in the suggestion breaks the detection loop because it checks topmost nodes (tree.body) (like if statement in this test case) in line_begins_import dict..

This is what my suggestion was meant to address: by using `nodes_of_class()` we visit every line instead of top-level nodes only, and later the `get()` call has a fallback to False for the nodes not stored in the dict. Or, you could also very simply adjust my suggestion to provide values for all lines.

> I think, the test case test_no_hide_code_with_imports can't be handled properly because existing code works on line-level not node-level. Therefore, all these lines do not start with imports and not skipped from similarity check.

I think we should skip them. Any line with an import anywhere in it should count as a skipped line. I think that's what my suggestion was doing.

***
Some test cases would be helpful to show this behavior. Thanks!

Reply:
Thank you for pointing out to `max-similarity-lines`, my test case was flawed.

I introduced and adapted your suggestion, which is definitely more flexible and concise.

> I think we should skip them. Any line with an import anywhere in it should count as a skipped line. I think that's what my suggestion was doing.

Good idea, adjusted this test case too.